libmamba/src/specs/build_number_spec.cpp

```diff
@@ -0,0 +1,331 @@
+            -> bool
+        {
+            return true;
+        }
+
+        auto
+        equal(std::less_equal<BuildNumberPredicate::BuildNumber>, std::less_equal<BuildNumberPredicate::BuildNumber>)
+            -> bool
+        {
+            return true;
+        }
+
+        template <typename U, typename... T>
+        auto equal_with(const std::variant<T...>& lhs, const std::variant<T...>& rhs) -> bool
+        {
+            return std::holds_alternative<U>(lhs) && std::holds_alternative<U>(rhs)
+                   && equal(std::get<U>(lhs), std::get<U>(rhs));
+        }
+
+        template <typename... T>
+        auto equal(const std::variant<T...>& lhs, const std::variant<T...>& rhs) -> bool
+        {
+            return (equal_with<T>(lhs, rhs) || ...);
+        }
+    }
+
+    /*****************************************
+     *  BuildNumberPredicate Implementation  *
+     *****************************************/
+
+    auto BuildNumberPredicate::contains(const BuildNumber& point) const -> bool
+    {
+        return std::visit([&](const auto& op) { return op(point, m_build_number); }, m_operator);
+    }
+
+    auto BuildNumberPredicate::make_free() -> BuildNumberPredicate
+    {
+        return BuildNumberPredicate({}, free_interval{});
+    }
+
+    auto BuildNumberPredicate::make_equal_to(BuildNumber ver) -> BuildNumberPredicate
+    {
+        return BuildNumberPredicate(std::move(ver), std::equal_to<BuildNumber>{});
+    }
+
+    auto BuildNumberPredicate::make_not_equal_to(BuildNumber ver) -> BuildNumberPredicate
+    {
+        return BuildNumberPredicate(std::move(ver), std::not_equal_to<BuildNumber>{});
+    }
+
+    auto BuildNumberPredicate::make_greater(BuildNumber ver) -> BuildNumberPredicate
+    {
+        return BuildNumberPredicate(std::move(ver), std::greater<BuildNumber>{});
+    }
+
+    auto BuildNumberPredicate::make_greater_equal(BuildNumber ver) -> BuildNumberPredicate
+    {
+        return BuildNumberPredicate(std::move(ver), std::greater_equal<BuildNumber>{});
+    }
+
+    auto BuildNumberPredicate::make_less(BuildNumber ver) -> BuildNumberPredicate
+    {
+        return BuildNumberPredicate(std::move(ver), std::less<BuildNumber>{});
+    }
+
+    auto BuildNumberPredicate::make_less_equal(BuildNumber ver) -> BuildNumberPredicate
+    {
+        return BuildNumberPredicate(std::move(ver), std::less_equal<BuildNumber>{});
+    }
+
+    auto BuildNumberPredicate::str() const -> std::string
+    {
+        return fmt::format("{}", *this);
+    }
+
+    BuildNumberPredicate::BuildNumberPredicate(BuildNumber ver, BinaryOperator op)
+        : m_build_number(std::move(ver))
+        , m_operator(std::move(op))
+    {
+    }
+
+    auto operator==(const BuildNumberPredicate& lhs, const BuildNumberPredicate& rhs) -> bool
+    {
+        return (lhs.m_build_number == rhs.m_build_number) && equal(lhs.m_operator, rhs.m_operator);
```

Comment:
Actually you could even perform the variant comparison here and get rid of the `  auto equal(const std::variant<T...>& lhs, const std::variant<T...>& rhs)` function.

Reply:
Again, I cannot specialized `operator==(std::equal_to<int>, std::equal_to<int>)` without UB.

Reply:
Is there a case where `operator==(std::equal_to<int>, std::equal_to<int>)` should return `false`? Otherwise, comparing the type is enough.